### PR TITLE
Switch back to `MutableMap` to avoid scatter map bug

### PR DESCRIPTION
**Pull request by JakeWharton**
Sometimes the scatter-based maps can return values from other keys when inserting a new key. See https://issuetracker.google.com/issues/378077858.

---

- [ ] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt [hunk 1]
```diff
@@ -235,9 +236,10 @@ public class HostProtocolAdapter<W : Any>(
       // Must have a Create node that precedes it.
       if (lastCreatedId != change.id) continue
 
-      idToNode[change.id.value] = ReuseNode(
-        widgetId = change.id,
-        childrenTag = ChildrenTag.Root,
+      idToNode[lastCreatedId.value] = ReuseNode(
+        widgetId = lastCreatedId,
```
Comment (JakeWharton) [inline redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt:241]: Switch to local which already has the same value as `change.id`

File: redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt [hunk 2]
```diff
@@ -359,19 +361,17 @@ public class HostProtocolAdapter<W : Any>(
      * descendants into the nodes map.
      */
     fun assignPooledNodeRecursive(
-      nodes: MutableIntObjectMap<ProtocolNode<W>>,
+      nodes: MutableMap<Int, ProtocolNode<W>>,
       changesAndNulls: Array<Change?>,
       pooled: ProtocolNode<W>,
     ) {
       // Reuse the node.
+      pooled.id = widgetId
```
Comment (JakeWharton) [inline redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt:370]: No need to lookup the `Create`, we already have a member with the new ID.

**Comment by JakeWharton** [comment]
Took a quick look around at other scatter-based map and set usage. They don't have this long-lived put/remove pattern.

File: redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt [hunk 3]
```diff
@@ -235,9 +236,10 @@ public class HostProtocolAdapter<W : Any>(
       // Must have a Create node that precedes it.
       if (lastCreatedId != change.id) continue
 
-      idToNode[change.id.value] = ReuseNode(
-        widgetId = change.id,
-        childrenTag = ChildrenTag.Root,
+      idToNode[lastCreatedId.value] = ReuseNode(
+        widgetId = lastCreatedId,
+        // This is the root of the reuse tree, so it will never have a children tag from a parent.
```
Comment (squarejesse) [inline redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt:242]: Thanks

**Review by squarejesse — APPROVED** [review]

**Pull request merged**